### DISCUSSION TITLE
Headphone out is way too soft

### DISCUSSION THREAD

**Discussion by greno** [pschatzmann/arduino-audiokit#62] [Q&A] [answered]
I have many of the samples working (player-sd-audiokit, streams-url_mp3-audiokit, audiokit-effects-audiokit), but I just can't increase the volume, no matter what I do.  With headphones, it's barely hearable.  With a powered speaker with the volume all the way up, it's still softer than a speaking voice.

My board is an AI Thinker ESP32 Audio Kit v2.2 A237.  I've changed the board to #5, #7 and #10, but only #5 and #10 work (but still low volume).  Is it possible this board is using some other chip set?

My goal is to eventually build a simple guitar effects box, using audiokit-effects-audiokit as a starting point.  But for now, I'm just trying to get a clean input and output signal with reasonable levels.  I've read through the thread with billy8119, who seems to have the same issue.

Here is what I have so far.  It's pretty simple.  If I set the mic gain above 18 db, I hear some clipping/distortion.  

// Audio
AudioKitStream kit; // Access I2S as stream
StreamCopy copier(kit, kit); // copy effects to kit

void setup() {
  Serial.begin(115200);
  delay(800);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  auto cfg = kit.defaultConfig(RXTX_MODE);
  cfg.sd_active = false;
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;

  // minimize lag
  cfg.buffer_count = 2;
  cfg.buffer_size = 256;

  kit.begin(cfg);
  kit.setVolume(.96);  
  es8388_set_mic_gain(MIC_GAIN_18DB);   
  
}

void loop() {
  copier.copy();

}

**Comment by pschatzmann** [comment]
For some strange reason only some Kits have issues with the volume:
Did you try the different AI_THINKER_ES8388_VOLUME_HACK settings ?

**Reply by greno** [reply]
Yes, I did and without luck. Setting to 0 made it even softer.  I couldn't tell the difference between 1 and 2.

Are other boards more reliable? Should I just get a LyraT board instead?  Or do you have any other recommendations?

**Comment by pschatzmann** [comment]
I was never using the headphone output, so I don't know.
Are you using earphones like the ones for a mobile phone ? 
Maybe you should select some with a lower impedence or you could try to optimize the CODEC register values for your requirements: https://datasheet.lcsc.com/lcsc/1912111437_Everest-semi-Everest-Semiconductor-ES8388_C365736.pdf

**Reply by greno** [reply]
Still troubleshooting.  What is odd is that when I hooked up a speaker to the speaker out, the line out volume jumped.  I'll try to recreate that.

**Reply by greno** [reply]
And now I can't recreate it.  At some point I had a clean, loud volume out of the line out. After unplugging and plugging in the speaker and the line out cable (not using headphones now), it's back to the low volume.  Something made it work for a while. I'll play around with the registers a bit.

**Reply by pschatzmann** [reply]
I would expect that the line out volume can be controlled via DAC Control 26 and DAC Control 27.
Looking at the code, AI_THINKER_ES8388_VOLUME_HACK has some logic updating the volume both for the amp and the aux out wherease the default implementation keeps them just at the default values.

**Comment by pschatzmann** [comment]
Just another idea: did you call kit.processActions() in the loop ?
This is determining (among others) if the headphones are plugged in and if they are, it deactivates the output via speakers. I could imagine that this might provide more power to the headphones...

In my last commit I also added the setSpeakerActive() method to the AudioKitStream(), so that you can deactivate the speaker explicitly w/o the key event handler

**Reply by greno** [reply]
I hadn't included the kit.processActions() call before, but adding it does what it's supposed to do - it cuts the speaker out when the headphones are plugged in.  But no change in volume.

I also created a new  AI_THINKER_ES8388_VOLUME_HACK==3 so I can play around with some of the registers.  Unlike hack 2, I set ES8388_DACCONTROL26 and ES8388_DACCONTROL27 to the same volume as ES8388_DACCONTROL25 and ES8388_DACCONTROL26, but no luck there.  I don't think that's it because 26 and 27 are just volumes for the second audio channel, which I don't think is being used.  And all 4 are just mixed down into the speaker and headphones out.

I've also tried playing with ES8388_DACCONTROL4 and ES8388_DACCONTROL5, but no luck there either.  They are already set to zero, which seems to be the max level.

**Reply by greno** [reply]
Seems like it might be a hardware problem.  But I can't figure out why I got good volume for a short time.  I have no idea now what I did, but it wasn't a code change. Maybe I made some physical connection while handling the board.

**Reply by pschatzmann** [reply]
I can confirm: if you do not want to use the power amplifier switch it off with

auto cfg = kit.defaultConfig(TX_MODE);
// sd_active is setting up SPI with the right SD pins by calling
// SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
cfg.speaker_active = false;
kit.begin(cfg);

Or by using the autiomatic determination using kit.processActions().

With no volume hack the AUX provides line level which is too low for headphones.
Only with the #define AI_THINKER_ES8388_VOLUME_HACK 1 you can control the volume with setVolume()

**Comment by pschatzmann** [comment, marked as answer]
I can confirm: if you do not want to use the power amplifier switch it off with 
<code>
  auto cfg = kit.defaultConfig(TX_MODE);
  // sd_active is setting up SPI with the right SD pins by calling 
  // SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);
  cfg.speaker_active = false;
  kit.begin(cfg);
</code>

With no volume hack the AUX provides line level which is too low for headphones.
Only with the #define AI_THINKER_ES8388_VOLUME_HACK 1 you can control the volume with setVolume()

The AI_THINKER_ES8388_VOLUME_HACK 2 is just setting the mic_gain to a different default value and has no impact because you overwrite the value when you call es8388_set_mic_gain()

**Comment by greno** [comment]
Success!  Setting volume hack to 1 along with whatever else I changed fixed the issue.  I'm getting a good clean sound from the line in to both the speaker and line out.  Thanks Phil!

I also tested the new cfg.speaker_active= false option, and can confirm it works.  

I think I read that with this board, you can't disable the built in microphone without disabling the line in, correct? I think the only solution is to physically modify the board?

Next up for me - start playing with the effects.  I'd like to add an EQ and reverb effect.  If I have questions with that, I'll post it in the audio tools discussion.

Thanks again Phil.
-Gregg
